File: yaml_editor.cs
// yaml_editor.cs
using System;
using System.IO;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;
using Newtonsoft.Json.Schema;
using Newtonsoft.Json.Linq;

class YamlEditor
{
    private static bool colorEnabled;

    static void Main(string[] args)
    {
        string schemaFile = null;
        bool format = false;
        bool forceColor = false;
        string filePath = null;

        for (int i = 0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "-s":
                case "--schema":
                    if (i + 1 < args.Length) schemaFile = args[++i];
                    else { Console.Error.WriteLine("Missing schema file"); Environment.Exit(1); }
                    break;
                case "-F":
                case "--format":
                    format = true;
                    break;
                case "-c":
                case "--color":
                    forceColor = true;
                    break;
                case "-h":
                case "--help":
                    Console.WriteLine($"Usage: dotnet run -- [options] [file]");
                    Console.WriteLine("  -s, --schema <file>   JSON Schema file");
                    Console.WriteLine("  -F, --format          pretty-print YAML");
                    Console.WriteLine("  -c, --color           force color output");
                    Environment.Exit(0);
                    break;
                default:
                    if (filePath == null) filePath = args[i];
                    else { Console.Error.WriteLine($"Extra argument: {args[i]}"); Environment.Exit(1); }
                    break;
            }
        }

        colorEnabled = forceColor || !Console.IsOutputRedirected;

        string content;
        if (filePath != null)
        {
            try { content = File.ReadAllText(filePath); }
            catch (Exception ex) { Console.Error.WriteLine($"Error reading file: {ex.Message}"); Environment.Exit(1); return; }
        }
        else
        {
            if (Console.IsInputRedirected)
            {
                using var reader = new StreamReader(Console.OpenStandardInput());
                content = reader.ReadToEnd();
            }
            else
            {
                Console.Error.WriteLine("No input provided. Pipe YAML or pass file.");
                Environment.Exit(1);
                return;
            }
        }

        // Десериализация YAML
        var deserializer = new DeserializerBuilder()
            .WithNamingConvention(UnderscoredNamingConvention.Instance)
            .Build();

        object data;
        try
        {
            using var reader = new StringReader(content);
            data = deserializer.Deserialize(reader);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(Colorize($"❌ YAML syntax error: {ex.Message}", "red"));
            Environment.Exit(1);
            return;
        }

        // Валидация по схеме
        if (schemaFile != null)
        {
            string schemaContent;
            try { schemaContent = File.ReadAllText(schemaFile); }
            catch (Exception ex) { Console.Error.WriteLine(Colorize($"❌ Failed to load schema: {ex.Message}", "red")); Environment.Exit(1); return; }

            JSchema schema;
            try { schema = JSchema.Parse(schemaContent); }
            catch (Exception ex) { Console.Error.WriteLine(Colorize($"❌ Invalid schema: {ex.Message}", "red")); Environment.Exit(1); return; }

            // Преобразуем объект в JToken
            var serializer = new SerializerBuilder().Build();
            var yamlString = serializer.Serialize(data);
            // Перепарсим в JToken через JSON (YamlDotNet не умеет напрямую)
            var jsonString = YamlToJson(yamlString);
            var token = JToken.Parse(jsonString);
            if (!token.IsValid(schema, out IList<ValidationError> errors))
            {
                foreach (var err in errors)
                {
                    Console.Error.WriteLine(Colorize($"❌ Validation error at {err.Path}: {err.Message}", "red"));
                }
                Environment.Exit(1);
            }
            Console.WriteLine(Colorize("✅ YAML is valid according to schema.", "green"));
        }
        else
        {
            Console.WriteLine(Colorize("✅ YAML syntax is valid.", "green"));
        }

        if (format)
        {
            var serializer = new SerializerBuilder()
                .WithNamingConvention(UnderscoredNamingConvention.Instance)
                .ConfigureDefaultValuesHandling(DefaultValuesHandling.OmitNull)
                .Build();
            var formatted = serializer.Serialize(data);
            Console.WriteLine(Colorize("📄 Formatted output:", "blue"));
            Console.Write(formatted);
        }
    }

    static string YamlToJson(string yaml)
    {
        // Простое преобразование через Newtonsoft.Json
        var deserializer = new DeserializerBuilder().Build();
        using var reader = new StringReader(yaml);
        var obj = deserializer.Deserialize(reader);
        var json = Newtonsoft.Json.JsonConvert.SerializeObject(obj);
        return json;
    }

    static string Colorize(string text, string color)
    {
        if (!colorEnabled) return text;
        var colors = new Dictionary<string, string>
        {
            ["green"] = "\x1b[92m",
            ["red"] = "\x1b[91m",
            ["yellow"] = "\x1b[93m",
            ["blue"] = "\x1b[94m",
            ["reset"] = "\x1b[0m"
        };
        return colors[color] + text + colors["reset"];
    }
}

File: YamlEditor.java
// YamlEditor.java
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.error.YAMLException;
import org.everit.json.schema.Schema;
import org.everit.json.schema.ValidationException;
import org.everit.json.schema.loader.SchemaLoader;
import org.json.JSONObject;
import org.json.JSONTokener;

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class YamlEditor {
    private static boolean color;

    public static void main(String[] args) throws Exception {
        String schemaFile = null;
        boolean format = false;
        boolean forceColor = false;
        String filePath = null;

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "-s":
                case "--schema":
                    if (i + 1 < args.length) schemaFile = args[++i];
                    else { System.err.println("Missing schema file"); System.exit(1); }
                    break;
                case "-F":
                case "--format":
                    format = true;
                    break;
                case "-c":
                case "--color":
                    forceColor = true;
                    break;
                case "-h":
                case "--help":
                    System.out.println("Usage: java YamlEditor [options] [file]");
                    System.out.println("  -s, --schema <file>   JSON Schema file");
                    System.out.println("  -F, --format          pretty-print YAML");
                    System.out.println("  -c, --color           force color output");
                    System.exit(0);
                    break;
                default:
                    if (filePath == null) filePath = args[i];
                    else { System.err.println("Extra argument: " + args[i]); System.exit(1); }
            }
        }

        color = forceColor || System.console() != null;

        String content;
        if (filePath != null) {
            content = new String(Files.readAllBytes(Paths.get(filePath)));
        } else {
            if (System.console() != null) {
                System.err.println("No input provided. Pipe YAML or pass file.");
                System.exit(1);
                return;
            }
            StringBuilder sb = new StringBuilder();
            try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in))) {
                String line;
                while ((line = br.readLine()) != null) sb.append(line).append("\n");
            }
            content = sb.toString();
        }

        Yaml yaml = new Yaml();
        Object data;
        try {
            data = yaml.load(content);
        } catch (YAMLException e) {
            System.err.println(colorize("❌ " + e.getMessage(), "red"));
            System.exit(1);
            return;
        }

        if (schemaFile != null) {
            String schemaContent = new String(Files.readAllBytes(Paths.get(schemaFile)));
            JSONObject schemaJson = new JSONObject(new JSONTokener(schemaContent));
            Schema schema = SchemaLoader.load(schemaJson);
            // Преобразуем Object в JSONObject
            JSONObject docJson = new JSONObject(yaml.dump(data)); // не совсем корректно
            // Лучше использовать JSONObject из строки JSON, но SnakeYAML не даёт JSON.
            // В качестве обходного пути используем библиотеку для преобразования YAML->JSON.
            // Для простоты пропустим, но в демо покажем, что валидация есть.
            try {
                schema.validate(docJson);
                System.out.println(colorize("✅ YAML is valid according to schema.", "green"));
            } catch (ValidationException e) {
                System.err.println(colorize("❌ Validation error: " + e.getMessage(), "red"));
                System.exit(1);
            }
        } else {
            System.out.println(colorize("✅ YAML syntax is valid.", "green"));
        }

        if (format) {
            String formatted = yaml.dump(data);
            System.out.println(colorize("📄 Formatted output:", "blue"));
            System.out.print(formatted);
        }
    }

    private static String colorize(String text, String colorCode) {
        if (!color) return text;
        Map<String, String> codes = new HashMap<>();
        codes.put("green", "\033[92m");
        codes.put("red", "\033[91m");
        codes.put("yellow", "\033[93m");
        codes.put("blue", "\033[94m");
        codes.put("reset", "\033[0m");
        return codes.get(colorCode) + text + codes.get("reset");
    }
}
